File: app/src/main/java/com/example/lawson/androidsummery/draw/DrawView.java
package com.example.lawson.androidsummery.draw;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewTreeObserver;

import com.example.lawson.androidsummery.R;

/**
 * Created by Ian.Lu on 2016/11/7.
 * Project : AndroidSummary
 */

public class DrawView extends View {

    private Canvas drawCanvas;
    private Bitmap drawBitmap;
    private Paint mPaint;
    private Path mPath;
    private float startX;
    private float startY;

    public DrawView(Context context) {
        super(context);
        init();
    }

    public DrawView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public DrawView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
            @Override
            public void onGlobalLayout() {
                Bitmap.Config config = Bitmap.Config.ARGB_4444;
                drawBitmap = Bitmap.createBitmap(getMeasuredWidth(), getMeasuredHeight(), config);

                drawCanvas = new Canvas(drawBitmap);

                mPaint = new Paint();
                mPaint.setAntiAlias(true);
                setColor(getContext().getResources().getColor(R.color.app_day_bg_color));
//                mPaint.setColor(Color.BLACK);
                mPaint.setStyle(Paint.Style.STROKE);
                mPaint.setStrokeWidth(2);

                mPath = new Path();

            }
        });

    }

    public void setColor(int color) {
        int red = (color & 0xff0000) >> 16;
        int green = (color & 0x00ff00) >> 8;
        int blue = (color & 0x0000ff);
        mPaint.setColor(Color.rgb(red, green, blue));
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        canvas.drawBitmap(drawBitmap, 0, 0, null);
        canvas.drawPath(mPath, mPaint);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        int action = event.getAction();
        float x = event.getX();
        float y = event.getY();

        switch (action) {
            case MotionEvent.ACTION_DOWN:
                mPath.moveTo(x, y);
                break;

            case MotionEvent.ACTION_MOVE:
                mPath.quadTo(startX, startY, x, y);
                break;

            case MotionEvent.ACTION_UP:
                drawCanvas.drawPath(mPath, mPaint);
                break;
        }

        invalidate();
        startX = x;
        startY = y;

        return true;
    }

    public Bitmap getDrawBitmap() {
        return drawBitmap;
    }
}

File: app/src/main/java/com/example/lawson/androidsummery/diyview/bord/view/DrawView.kt
package com.example.lawson.androidsummery.diyview.bord.view

import android.app.Activity
import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.util.Log
import android.view.MotionEvent
import android.view.View
import com.example.lawson.androidsummery.diyview.MeasureUtil

/**
 * USER：lujianzhi
 * DATE：2020/10/26
 */
class DrawView @JvmOverloads constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int = 0) :
        View(context, attrs, defStyleAttr) {

    private val mPaint = Paint()
    private val mPath = Path()
    private var mCanvas: Canvas? = null
    private lateinit var mBitmap: Bitmap
    private var mInitBitmap: Bitmap? = null

    private var preX: Float = 0.0f
    private var preY: Float = 0.0f

    init {
        mPaint.strokeWidth = 10.0F
        mPaint.color = Color.parseColor("#FF0000")
        mPaint.style = Paint.Style.STROKE
        mPaint.strokeJoin = Paint.Join.ROUND
        mPaint.isAntiAlias = true
    }

    fun saveLayer(){
    }

    fun eraser() {
        mPaint.xfermode = PorterDuffXfermode(PorterDuff.Mode.CLEAR)
    }

    fun paint() {
        mPaint.xfermode = null
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)

        if (mInitBitmap == null || mCanvas == null) {
            mBitmap = Bitmap.createBitmap(measuredWidth, measuredHeight, Bitmap.Config.ARGB_4444)
            mCanvas = Canvas(mBitmap)
            mInitBitmap = mBitmap
        }
    }

    override fun onDraw(canvas: Canvas?) {
        super.onDraw(canvas)
        canvas?.drawBitmap(mBitmap, 0F, 0F, null)
    }

    override fun onTouchEvent(event: MotionEvent?): Boolean {
        val x: Float = event?.x ?: 0F
        val y: Float = event?.y ?: 0F

        when (event?.action) {
            MotionEvent.ACTION_DOWN -> {
                mPath.moveTo(x, y)
            }
            MotionEvent.ACTION_MOVE -> {
                //贝塞尔曲线，更加圆滑
                mPath.quadTo(preX, preY, x, y)
                mCanvas?.drawPath(mPath, mPaint)
            }
            MotionEvent.ACTION_UP -> {
                mPath.reset()
            }
            else -> {

            }
        }

        invalidate()
        preX = x
        preY = y
        return true
    }
}